Meta: Add feature option search

Mottie opened the pull request:
## Description

This PR adds a feature option search to Refined GitHub's option popup.

Searching for `open batch` will only return one result ("batch-open-issues"); the order of the words doesn't matter. Under the hood, this makes sure that every word in the search pattern matches some part of the feature name and description.

Notes:

- This only searches the feature name & description.
- All matches are case-insensitive.
- The following characters are ignored: <code>,.:;'"’`</code>.
- The following character is replaced with a space: `-`; this allows `"open-batch` to still match "batch-open-issues"

## Screenshots

![search](https://user-images.githubusercontent.com/136959/68098648-78a8a580-fe83-11e9-9f97-bcaa2a830787.gif)

## Closes
Closes #2147

## Test

Open Refined-GitHub's options

yakov116 commented on source/options.tsx:
```diff
@@ -48,6 +49,53 @@ function buildFeatureCheckbox({name, description, screenshot, disabled}: Feature
+		const pattern = (event.target as HTMLInputElement).value || '';
+		const patternArray = replaceCharacters(pattern)
+			.split(separators)
+			.filter(s => s); // Remove empty strings
```
```suggestion
			.filter(Boolean); // Remove empty strings
```

fregante commented:
Great to see this!

I don't think one ever needs the "non-exact" search, which seems to me it's just an "keyword OR keyword" search. If I enter more words, I want all of them to appear.

For the implementation, have a look at this, it's shorter:

https://github.com/HangYang/one-click-extensions-manager/blob/1ac4566f7d28329c98ce8f0edc7a380a8ce9ecf7/src/popup.js#L94-L102

Mottie commented:
I removed the basic match and rebased from the master. This made my screenshot dark 😺

Oh, you've updated your comment with code... I replaced some of the punctuation in the search because it's not so easy to type `’` on some computers.

Mottie commented:
Ok, I've updated the code. The search from the screenshot doesn't work the same because there are no longer any string replacements.

fregante commented on source/options.tsx:
```diff
@@ -62,6 +62,17 @@ async function init(): Promise<void> {
+		for (const feature of select.all('.feature')) {
+			feature.hidden = hasKeywords ?
+				!keywords.every(word => feature.dataset.text.includes(word)) :
+				false;
```
This shouldn't be necessary. I _think_ this is equal:

```js
feature.hidden = !keywords.every(word => feature.dataset.text.includes(word));
```

Mottie replied:
You're right... like I said my computer was doing some crazy stuff. I cleared the input and all the features stayed hidden. I thought I was going crazy. That and the times the input would enter text in backwards.

fregante commented on source/options.css:
```diff
@@ -110,7 +110,8 @@ h2 ~ h2 {
 
-.js-features :not(:checked) + .info .extended-options {
+.js-features :not(:checked) + .info .extended-options,
+.feature[hidden] {
```
This shouldn't be needed, `[hidden] {display:none}` is offered by the browser

Mottie replied:
It doesn't seem to be set... having the hidden attribute was not hiding the entries.

fregante commented:
Also everything should be renamed to `filter` rather than `search` https://ux.stackexchange.com/a/4756/35104

Mottie commented:
Umm, please go ahead and make these changes... I'm having some issues with my computer.

Apparently running `npm install` now makes my computer reboot.

Then I saw this:

<img width="483" alt="2019-11-03 22_46_20-MINGW64__x_Repos_+Misc_refined-github" src="https://user-images.githubusercontent.com/136959/68100217-6ed77000-fe8c-11e9-9ab8-77f429e87f1a.png">

And after a repair:

<img width="115" alt="2019-11-03 22_51_20-Settings" src="https://user-images.githubusercontent.com/136959/68100231-857dc700-fe8c-11e9-8c6a-1dba4e2bdffa.png">

I think I'm gonna call it a night LOL

Mottie commented:
So... `npm install` is still having issues, but apparently `yarn install` worked.

Then YAY, I made the changes.

`npm test` failed and was showing some weird error message. `yarn test` caused a GSOD (apparently Windows insider now shows a green-screen of death vs the older, outdated blue-screen of death).

Tests seem to be passing here (finally) LOL... I still don't know how to deal with the `feature.dataset.text` TS error other than wrapping it in a `String()`.

_Now_, I'm going to bed.

Mottie commented:
And oops... I removed the `[hidden] { display: none; }` so now the features aren't getting hidden.

fregante commented:
> And oops... I removed the `[hidden] { display: none; }` so now the features aren't getting hidden.

Uhm, what browser are you using? It works correctly in both Firefox and Chrome for me

Mottie commented:
👍 Thanks for the fixes! Typescript is still weird to me.

jerone commented:
Great work @Mottie 👍
